src/main/scala/lms/thirdparty/cuda/cuda.scala

```diff
@@ -254,6 +269,65 @@ object CUDATypeLess extends Dsl with StackArrayOps with CLibs with CudaFunction
     val kernel = CUDA_DIV_KERNEL(a.et)
     kernel(a, b, res, size, DIM3(gridSize), DIM3(blockSize))
   }
+
+  // This split op kernel has lots of limitations
+  // 1. the input (0th parameter) is split to 2 outputs (1st and 2nd parameter)
+  // 2. the input is 2D, and the split axis is 1
+  // 3. the split is by equal sizes, so the input dim1 size must be even number
+  // 4. the 3rd input is size of dim 0 of input, the 4th input is 1/2 of size of dim 1 of input
+  def CUDA_SPLIT2_2D1_EQUAL_KERNEL(m: Manifest[_], comment: String = "")(implicit __pos: SourceContext) = CUDA_KERNEL5({xn: List[Backend.Exp] =>
+    withComment(comment) {
+      // type cast
+      val input = (new ARRAY(xn(0))).withSrcType(__pos, m.arrayManifest)
+      val output0 = (new ARRAY(xn(1))).withSrcType(__pos, m.arrayManifest)
+      val output1 = (new ARRAY(xn(2))).withSrcType(__pos, m.arrayManifest)
+      val dim0 = (new INT(xn(3))).withSrcType(__pos, manifest[Int])
+      val dim1 = (new INT(xn(4))).withSrcType(__pos, manifest[Int])
+
+      // actual computation
+      val stride = gridDimX * blockDimX
+      val tid = threadIdxX + blockIdxX * blockDimX
+      for (i <- range_until_step(Wrap[Int](tid.x), Wrap[Int]((dim0*dim1).x), Wrap[Int](stride.x))) {
+        val index = INT(Unwrap(i))
+        val size0 = index / dim0
+        val size1 = index % dim0
+        __ifThenElse(Wrap[Boolean]((size1 < dim1).x),
+          { output0(INT(size0 * dim1 + size1)) = input(index); () },
+          { output1(INT(size0 * dim1 + size1 - dim1)) = input(index); () })
+      }
+      Backend.Const(())
```

Comment:
The mixed use of "untyped" interface and "typed" interface as well as `__ifThenElse` looks quite cumbersome.

Reply:
I know :)

Tiark said "using typeless API lazily" (i.e. use them when it is necessary)

Reply:
__ifThenElse is necessary because we cannot use micros here (i.e., we cannot use `@virtualize`)